### ISSUE TITLE
BUG: IPX now reports 0 crossover iterations

### ISSUE THREAD

**Issue by mckib2** [ERGO-Code/HiGHS#473]
There appears to be a regression in how crossover iterations are reported for IPX corresponding to [this SciPy unit test](https://github.com/scipy/scipy/blob/b986aaaa3b9a2b2c2008db43db53ca5d2ab47e24/scipy/optimize/tests/test_linprog.py#L1632).

The following simple problem should have non-zero crossover iterations:
[crossover.mps.txt](https://github.com/ERGO-Code/HiGHS/files/6017257/crossover.mps.txt)

Command: `./highs --solver ipm crossover.mps`

Here is the console output showing a single primal push:
```
Running HiGHS 1.0.0 [date: 2021-02-17, git hash: ed3a64f58]
Copyright (c) 2020 ERGO-Code under MIT licence terms

LP       : crossover
Rows     : 1
Cols     : 2
Nonzeros : 2
Presolve : Reductions: rows 1(-0); columns 2(-0); elements 2(-0)
INFO   : Solving the presolved LP
INFO   : IPX model has 1 rows, 2 columns and 2 nonzeros
Input
    Number of variables:                                2
    Number of free variables:                           0
    Number of constraints:                              1
    Number of equality constraints:                     0
    Number of matrix entries:                           2
    Matrix range:                                       [1e+00, 1e+00]
    RHS range:                                          [1e+00, 1e+00]
    Objective range:                                    [1e+00, 1e+00]
    Bounds range:                                       [1e+00, 1e+00]
Preprocessing
    Dualized model:                                     no
    Number of dense columns:                            0
    Range of scaling factors:                           [1.00e+00, 1.00e+00]
IPX version 1.0
Interior Point Solve
 Iter     P.res    D.res            P.obj           D.obj        mu     Time
   0   1.71e+00 1.92e+00  -6.66666667e-01 -4.84005365e+00  4.46e+00       0s
   1   1.71e-06 1.92e-06  -6.13119184e-01 -3.44211579e+00  5.66e-01       0s
   2   1.71e-12 2.53e-07  -8.83385521e-01 -1.23585598e+00  7.05e-02       0s
   3   5.55e-17 1.55e-09  -9.98858659e-01 -1.00163518e+00  5.55e-04       0s
   4*  4.55e-18 1.55e-15  -9.99999999e-01 -1.00000000e+00  5.55e-10       0s
 Constructing starting basis...
Crossover
    Primal residual before push phase:                  1.32e-09
    Dual residual before push phase:                    6.18e-10
    Number of dual pushes required:                     0
    Number of primal pushes required:                   1
Summary
    Runtime:                                            0.00s
    Status interior point solve:                        optimal
    Status crossover:                                   optimal
    objective value:                                    -9.99999999e-01
    interior solution primal residual (abs/rel):        0.00e+00 / 0.00e+00
    interior solution dual residual (abs/rel):          1.33e-15 / 6.66e-16
    interior solution objective gap (abs/rel):          2.78e-09 / 1.39e-09
    basic solution primal infeasibility:                0.00e+00
    basic solution dual infeasibility:                  0.00e+00
INFO   : Ipx: IPM       optimal
INFO   : Ipx: Crossover optimal
INFO   : Solving the original LP from the solution after postsolve
INFO   : Scaling: Matrix has [min, max] values of [1, 1] within [0.2, 5] so no scaling performed
INFO   : Basis condition estimate of           1 is within the tolerance of 1e+14
Postsolve  : 0
Time       :     0.03
Time Pre   :     0.00
Time PreLP :     0.03
Time PostLP:     0.00
For LP        crossover: Presolve     0.00 (  2%): Solve presolved LP     0.03 ( 94%): Solve original LP     0.00 (  1%)

Model   status      : Optimal
Primal  status      : Feasible point
Dual    status      : Feasible point
Simplex   iterations: 0
IPM       iterations: 4
Objective value     : -1.0000000000e+00
HiGHS run time      :          0.03
```

**Comment by jajhall** [comment]
The changes I made to IPX yesterday, which included the change I made corresponding to #481, have been objected to by the person who maintains the IPX presence in HiGHS. We need to keep differences between IPX in https://github.com/ERGO-Code/HiGHS and https://github.com/ERGO-Code/ipx as small as possible, particularly in the area of crossover. 

Hence I've reverted my changes, so IPX will report 0 crossover iterations in this case. 

TBH I don't know when/why we switched from counting basis updates to pushes in IPX crossover. Perhaps "pushes" includes all basis updates but, if not, they underestimate the cost of crossover since - as I understand - a "push" is a change of variables that doesn't involve a change of basis.

For me this isn't a big enough issue to justify further work

**Comment by mckib2** [comment]
@jajhall I understand the concern about keeping the repositories in sync.  I am concerned about the build updates that were needed to build on Windows, macOS, etc.  Will we be able to keep those updates (not crossover related)?  If not, we may not be able to include IPX in SciPy or we may need to rely on an older version that has these updates

**Comment by mckib2** [comment]
We can also propose these changes to https://github.com/ERGO-Code/ipx to make sure we're keeping both aligned if you think Lukas would be responsive to that

**Comment by jajhall** [comment]
> @jajhall I understand the concern about keeping the repositories in sync. I am concerned about the build updates that were needed to build on Windows, macOS, etc. Will we be able to keep those updates (not crossover related)? If not, we may not be able to include IPX in SciPy or we may need to rely on an older version that has these updates.
> 
> In principle it's fine to count basis updates instead of pushes -- we had just noticed that the behavior changed. If this is a reasonable thing to do, then we can update our tests accordingly

What updates were needed to build on Windows, macOS, etc? 

Other than the "crossover" count, little has changed with IPX functionally. It's inevitable that things like counts change as code is developed, anyway. Tests based on counts simply have to be updated. We have to do it with our tests. The only tests that wouldn't change are those based on things like problem status and objective function value.

IPX isn't independent of the rest of HiGHS, so you can't use an older version unless you use an older version of the whole of HiGHS. Older versions of HiGHS won't be supported or documented. Also, you have issues relating to primal simplex that wouldn't be addresses. And you wouldn't get the MIP solver.

> We can also propose these changes to https://github.com/ERGO-Code/ipx to make sure we're keeping both aligned if you think Lukas would be responsive to that

Lukas wouldn't be responsive to that. I can ask him what he thinks the proper "crossover count" should be, though.

**Comment by mckib2** [comment]
> What updates were needed to build on Windows, macOS, etc?

I think the necessary changes were included in #491, so all that's left for me is to verify the build configurations.  We were experiencing some strange seg faults using IPX on macOS but I suspect these have been resolved (or there might be issues in our wrapper that we need to debug)

> IPX isn't independent of the rest of HiGHS, so you can't use an older version unless you use an older version of the whole of HiGHS. Older versions of HiGHS won't be supported or documented. Also, you have issues relating to primal simplex that wouldn't be addressed. And you wouldn't get the MIP solver.

I think we're all trying to do the same thing -- make minimal changes to sources we don't own!  We have been able to use HiGHS without any SciPy-specific changes so far and I'm hoping to continue like this.  I appreciate you being responsive these build issues.  We are in the near future looking at including more sensitivity analysis that you've already seen and HiGHS MIP solving capabilities
